validator_client/src/config.rs

```diff
@@ -1,32 +1,131 @@
+    /// The chain specification that we are connecting to
     pub spec: ChainSpec,
 }
 
-const DEFAULT_LIGHTHOUSE_DIR: &str = ".lighthouse-validators";
+const DEFAULT_VALIDATOR_DATADIR: &str = ".lighthouse-validator";
+const DEFAULT_VALIDATORS_SUBDIR: &str = "validators";
+const DEFAULT_PRIVATE_KEY_FILENAME: &str = "private.key";
 
-impl ClientConfig {
-    /// Build a new configuration from defaults.
-    pub fn default() -> Self {
-        let data_dir = {
-            let home = dirs::home_dir().expect("Unable to determine home dir.");
-            home.join(DEFAULT_LIGHTHOUSE_DIR)
+impl ValidatorClientConfig {
+    /// Build a new configuration from defaults, which are overrided by arguments provided.
+    pub fn build_config(arguments: &ArgMatches) -> Result<Self, Error> {
+        // Use the specified datadir, or default in the home directory
+        let data_dir: PathBuf = match arguments.value_of("datadir") {
+            Some(path) => PathBuf::from(path.to_string()),
+            None => {
+                let home = dirs::home_dir().ok_or_else(|| {
+                    Error::new(ErrorKind::NotFound, "Unable to determine home directory.")
+                })?;
+                home.join(DEFAULT_VALIDATOR_DATADIR)
+            }
         };
-        fs::create_dir_all(&data_dir)
-            .unwrap_or_else(|_| panic!("Unable to create {:?}", &data_dir));
-        let server = "localhost:50051".to_string();
-        let spec = ChainSpec::foundation();
-        Self {
+        fs::create_dir_all(&data_dir)?;
+
+        let validator_dir = data_dir.join(DEFAULT_VALIDATORS_SUBDIR);
+        fs::create_dir_all(&validator_dir)?;
+
+        let server: String = match arguments.value_of("server") {
+            Some(srv) => {
+                //TODO: I don't think this parses correctly a server & port combo
+                srv.parse::<u16>()
+                    .map_err(|e| Error::new(ErrorKind::InvalidInput, e))?
+                    .to_string()
+            }
+            None => "localhost:50051".to_string(),
+        };
+
+        // TODO: Permit loading a custom spec from file.
+        let spec: ChainSpec = match arguments.value_of("spec") {
+            Some(spec_str) => {
+                match spec_str {
+                    "foundation" => ChainSpec::foundation(),
+                    "few_validators" => ChainSpec::few_validators(),
+                    // Should be impossible due to clap's `possible_values(..)` function.
+                    _ => unreachable!(),
+                }
+            }
+            None => ChainSpec::foundation(),
+        };
+
+        Ok(Self {
             data_dir,
+            validator_dir,
             server,
             spec,
+        })
+    }
+
+    /// Try to load keys from validator_dir, returning None if none are found or an error.
+    pub fn fetch_keys(&self) -> Result<Option<Vec<Keypair>>, Error> {
+        let mut validator_dirs = fs::read_dir(&self.validator_dir)?.peekable();
```

Comment:
I wouldn't use `peekable()` here. It adds a bit of complexity and latter down you are unwrapping because of it (line 88). 
There is varying ways to do this. The easiest to read would be something like this: 
```rust 
 let mut key_pairs: Vec<Keypair> = Vec::new();
 for validator_dir in fs::read_dir(&self.validator_dir)? {
            // Try to open the key file directly
            // TODO skip keyfiles that are not found, and log the error instead of returning it.
            let mut key_file = File::open(validator_dir.path().join(DEFAULT_PRIVATE_KEY_FILENAME))?;

            let key: Keypair = bincode::deserialize_from(&mut key_file)
                .map_err(|e| Error::new(ErrorKind::InvalidData, e))?;
            ...
}
```

or a more rust idomatic way, would be something like
```rust
let keypairs = fs::read_dir(&self.validator_dir)?.map(|validator_dir| {
     let key: Keypair = bincode::deserialize_from(&mut key_file)
                .map_err(|e| Error::new(ErrorKind::InvalidData, e))?;

            // TODO skip keyfile if it's not matched, and log the error instead of returning it.
            let validator_directory_name =
                validator_dir.file_name().into_string().map_err(|_| {
                    Error::new(
                        ErrorKind::InvalidData,
                        "The filename cannot be parsed to a string.",
                    )
                })?;
            if key.identifier() != validator_directory_name {
                return Err(Error::new(
                    ErrorKind::InvalidData,
                    "The validator directory ID did not match the key found inside.",
                ));
            }
            key }).collect();
            if key_pairs.is_none() { return Ok(None)}
            Ok(Some(keypairs))
}

Reply:
Typically if you are building an array from an iterator, its quite handy to `map().collect()`. Saves the initial `mut Vec::new()`.

Reply:
I'm not sure that @AgeManning's example here will work, because he's trying to use `?` inside a `map` that should collect to `Keypair` not `Result<Keypair, _)>`.

For example,

```rust
let cats: Vec<Cat> = kittens.map(|k| k.grow_up()?).collect();
```

This will not compile because `k.grow_up()?` is trying to return `Result<Cat, GrowError>` and `.collect()` is expecting a vec of `Cat`.

For handing errors during `Iterator` ops, you have [`try_fold`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.try_fold) and [`try_for_each`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.try_for_each). These will "short-circut" and stop as soon as one of the closures returns `Err`.

There's no `try_map` for a reason I don't yet understand. There is `filter_map` but that doesn't stop as soon as it reaches an `Err`. Instead it just filters things that aren't `Some` (it deals with `Option` not `Result`). WRT `try_map`, you could probably compose your own by doing a map to `Result` then `try_fold` into a vec. Sounds pretty awkward and slow, though.

IMO, what you have here is good.

Reply:
Thanks a lot for all the input.
I ended up using `filter_map` and `.ok()?`